internal/handler/control.go

```diff
@@ -1,22 +1,142 @@
 // -*- Mode: Go; indent-tabs-mode: t -*-
 //
 // Copyright (C) 2017-2018 Canonical Ltd
+// Copyright (C) 2020 IOTech Ltd
 //
 // SPDX-License-Identifier: Apache-2.0
 
 package handler
 
 import (
+	"context"
 	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"sync/atomic"
+	"time"
 
+	"github.com/edgexfoundry/device-sdk-go/internal/cache"
 	"github.com/edgexfoundry/device-sdk-go/internal/common"
+	"github.com/edgexfoundry/device-sdk-go/pkg/models"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+	"github.com/google/uuid"
 )
 
-func DiscoveryHandler(requestMap map[string]string) {
-	common.LoggingClient.Info(fmt.Sprintf("service: discovery request"))
-}
+var (
+	id     string
+	locker uint32
+)
 
 func TransformHandler(requestMap map[string]string) (map[string]string, common.AppError) {
 	common.LoggingClient.Info(fmt.Sprintf("service: transform request: transformData: %s", requestMap["transformData"]))
 	return requestMap, nil
 }
+
+func DiscoveryHandler(w http.ResponseWriter) {
+	if id == "" {
+		id = uuid.New().String()
+	}
+
+	if w != nil {
+		msg := fmt.Sprintf("Discovery triggered or already running, id = %s", id)
+		w.WriteHeader(http.StatusAccepted) //status=202
+		_, _ = io.WriteString(w, msg)
+	}
+
+	// locker is used to prevent duplicate call of protocol specific device discovery
+	// locker would be released once the DS main thread received the candidate devices
+	// found by discovery
+	if !atomic.CompareAndSwapUint32(&locker, 0, 1) {
+		common.LoggingClient.Info(fmt.Sprintf("Discovery request returned. discovery process is running"))
+		return
+	}
+	common.LoggingClient.Info(fmt.Sprintf("service %s discovery triggered", common.ServiceName))
+
+	ctx := context.WithValue(context.Background(), common.CorrelationHeader, id)
+	deviceCh := make(chan []models.DiscoveredDevice)
+	go common.Discovery.Discover(deviceCh)
+	go filterAndAddition(ctx, deviceCh)
+}
+
+func filterAndAddition(ctx context.Context, deviceCh <-chan []models.DiscoveredDevice) {
+	pws := cache.ProvisionWatchers().All()
+	devices := <-deviceCh
```

Comment:
The current implementation expects a single slice of discovered devices to be read from the device channel as a result of the call to `Discover`. The design actually says:

> Note: The implementation should call back for every device found.

and later...

> The filtered device addition function will take as an argument a collection of structs containing the above data. An implementation may choose to make one call per discovered device, but implementors are encouraged to batch the devices if practical, as in future EdgeX versions it will be possible for the SDK to create all required new devices in a single call to core-metadata.
> 
> Rationale: An alternative design would have the implementation function return the collection of discovered devices to the SDK.
>  * Allows for asynchronous operation. ..
>  * Allows DS implementations where devices self-announce to call the filtered device addition function independent of the discovery process

From my interpretation of the design, I think the receiver (ie. filterAndAdd) could possible read multiple slices from the implementation, and that the implementation should indicate somehow when it's done (i.e. by sending an empty slice or closing the channel).

@iain-anderson can you please review, and let's discuss Mon?

Reply:
Suggest a goroutine to process devices from the channel, to run for the lifetime of the service, as per processAsyncResults()